pandas/core/indexes/base.py

```diff
@@ -3317,7 +3317,7 @@ def _join_monotonic(self, other, how='left', return_indexers=False):
 
     def _wrap_joined_index(self, joined, other):
         name = self.name if self.name == other.name else None
-        return Index(joined, name=name)
+        return self.__class__(joined, name=name)
```

Comment:
Yes, good catch. But can you do this in a separate PR? As this will need some more fixes and more tests (eg for the case where both are not exactly the same)

In this PR you can just for now skip the test in case the index is a categorical index